### PR TITLE
Making it possible to pass Windows credential specs directly to the engine

### PR DESCRIPTION
**Pull request by wk8**
Instead of having to go through files or registry values as is currently the
case.

While adding GMSA support to Kubernetes (https://github.com/kubernetes/kubernetes/pull/73726)
I stumbled upon the fact that Docker currently only allows passing Windows
credential specs through files or registry values, forcing the Kubelet
to perform a rather awkward dance of writing-then-deleting to either the
disk or the registry to be able to create a Windows container with cred
specs.

This patch solves this problem by making it possible to directly pass
whole base64-encoded cred specs to the engine's API. I took the opportunity
to slightly refactor the method responsible for Windows cred spec as it
seemed hard to read to me.

Added some unit tests on Windows credential specs handling, as there were
previously none.

Added/amended the relevant integration tests.

I have also tested it manually: given a Windows container using a cred spec
that you would normally start with e.g.
```powershell
docker run --rm --security-opt "credentialspec=file://win.json" mcr.microsoft.com/windows/servercore:ltsc2019 nltest /parentdomain
# output:
# my.ad.domain.com. (1)
# The command completed successfully
```
can now equivalently be started with
```powershell
$rawCredSpec = & cat 'C:\ProgramData\docker\credentialspecs\win.json'
$escaped = $rawCredSpec.Replace('"', '\"')
docker run --rm --security-opt "credentialspec=raw://$escaped" mcr.microsoft.com/windows/servercore:ltsc2019 nltest /parentdomain
# same output!
```

I'll do another PR on Swarmkit after this is merged to allow services to use
the same option.

(It's worth noting that @dperny faced the same problem adding GMSA support
to Swarmkit, to which he came up with an interesting solution - see
https://github.com/moby/moby/pull/38632 - but alas these tricks are not
available to the Kubelet.)

### CODE REVIEW

File: daemon/oci_windows.go
Comment (thaJeztah) [inline daemon/oci_windows.go]: wondering (haven't fully thought this through); should we remove the `config` option, and make swarmkit convert it to `base64` (read the credential-spec from the config, and pass it as base64)?

@dperny @cpuguy83

File: daemon/oci_windows.go
Comment (thaJeztah) [inline daemon/oci_windows.go]: I saw @dperny was working on some of those, but while we're changing this, we should change these to;

```suggestion
		return "", errors.Wrapf(err, "credential spec for container %s could not be read from file %q", id, full)
```

File: daemon/oci_windows.go
Comment (thaJeztah) [inline daemon/oci_windows.go]: Same here

```suggestion
		return "", errors.Wrapf(err, "error reading credential spec %q from registry for container %s", name, id)
```

File: daemon/oci_windows.go
Comment (thaJeztah) [inline daemon/oci_windows.go]:
```suggestion
		return "", errors.Wrapf(err, "failed handling spec %q for container %s - registry key %s could not be opened", name, id, credentialSpecRegistryLocation)
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: daemon/oci_windows.go
Comment (wk8) [reply daemon/oci_windows.go]: I had the same question, actually :)

File: daemon/oci_windows.go
Comment (wk8) [reply daemon/oci_windows.go]: Good call, doing :)

File: daemon/oci_windows.go
Comment (cpuguy83) [reply daemon/oci_windows.go]: Yes, please ❤️

Comment (wk8) [reply daemon/oci_windows.go]: Happy to make a PR for that on Swarmkit after this is merged is @dperny likes the idea too :)

Comment (cpuguy83) [reply daemon/oci_windows.go]: Does it require a swarmkit change? I think the node agent can just read the config and stick it into the the credential spec?

File: daemon/oci_windows.go
Comment (wk8) [reply daemon/oci_windows.go]: 👍

File: daemon/oci_windows.go
Comment (wk8) [reply daemon/oci_windows.go]: 👍

File: daemon/oci_windows.go
Comment (wk8) [reply daemon/oci_windows.go]: @cpuguy83 : you may very well be right, I didn't look into it before opening my big mouth, much to my shame. I just assumed that it did require a swarmkit change.

Comment (wk8) [reply daemon/oci_windows.go]: Would still very much like @dperny 's take on this though before I do make that change :)

Comment (dperny) [reply daemon/oci_windows.go]: It would theoretically be a bit of a pain in the butt, but we could do that. You'd basically have to read the CredentialSpec at some point when converting to a container, and write it to base64. It should be done all in the engine, no swarmkit changes necessary.

File: daemon/oci_windows.go
Comment (dperny) [inline daemon/oci_windows.go]: I have a PR in the works (#38770) to normalize all of the error values to use the `errdefs` definitions. Can you modify this and any other _new_ errors to use those error types, so that I don't have to repeat the work when rebasing that PR?

File: daemon/oci_windows.go
Comment (thaJeztah) [reply daemon/oci_windows.go]: @wk8 just thinking of this; the credentials-spec file is always JSON, correct? Do we need to pass it as base64, or should we just pass as a plain string? (optionally, validating for well-formedness)

**Comment by thaJeztah** [comment]
Looks like there's failures on Windows;

```
01:50:37 FAIL: docker_cli_run_test.go:4146: DockerSuite.TestRunCredentialSpecFailures
01:50:37 
01:50:37 docker_cli_run_test.go:4159:
01:50:37     c.Assert(err.Error(), checker.Contains, attempt.expectedError, check.Commentf("%s expected %s got %s", attempt.value, attempt.expectedError, err))
01:50:37 ... obtained string = "" +
01:50:37 ...     "\n" +
01:50:37 ...     "Command:  d:\\CI\\CI-34c625c5ff\\binary\\docker.exe run --security-opt=credentialspec=rubbish busybox true\n" +
01:50:37 ...     "ExitCode: 125\n" +
01:50:37 ...     "Error:    exit status 125\n" +
01:50:37 ...     "Stdout:   \n" +
01:50:37 ...     "Stderr:   d:\\CI\\CI-34c625c5ff\\binary\\docker.exe: error during connect: Post http://127.0.0.1:2357/v1.30/containers/7e999460d094c24e4ebea2e71c3f8e33cdfd700ccfe3a476766cb6886ceff284/start: EOF.\n" +
01:50:37 ...     "\n" +
01:50:37 ...     "\n" +
01:50:37 ...     "Failures:\n" +
01:50:37 ...     "ExitCode was 125 expected 0\n" +
01:50:37 ...     "Expected no error"
01:50:37 ... substring string = "invalid credential spec security option - value must be prefixed file:// or registry://"
01:50:37 ... rubbish expected invalid credential spec security option - value must be prefixed file:// or registry:// got 
01:50:37 Command:  d:\CI\CI-34c625c5ff\binary\docker.exe run --security-opt=credentialspec=rubbish busybox true
01:50:37 ExitCode: 125
01:50:37 Error:    exit status 125
01:50:37 Stdout:   
01:50:37 Stderr:   d:\CI\CI-34c625c5ff\binary\docker.exe: error during connect: Post http://127.0.0.1:2357/v1.30/containers/7e999460d094c24e4ebea2e71c3f8e33cdfd700ccfe3a476766cb6886ceff284/start: EOF.
01:50:37 
01:50:37 
01:50:37 Failures:
01:50:37 ExitCode was 125 expected 0
01:50:37 Expected no error
```

File: daemon/oci_windows.go
Comment (wk8) [reply daemon/oci_windows.go]: Sure thing, doing that today or tomorrow.

**Comment by wk8** [comment]
@thaJeztah : thanks, having a look at the failure shortly.

File: daemon/oci_windows.go
Comment (wk8) [reply daemon/oci_windows.go]: Yes it is always JSON, so it could just be passed as a plain string.

In my experience though, that can fast become a pain with quote escaping when eg building a command line in a bash script, that's why I thought base64 might be a better idea? Happy to change though if you feel strongly about it.

JSON validation is an idea though. Didn't do it upfront because we don't seem to validate anything just yet - if we do validate, we should do it for all the different ways that exist; and that would arguably be somewhat of a breaking change. Happy to add it if you wish though.

File: daemon/oci_windows.go
Comment (wk8) [reply daemon/oci_windows.go]: Done :)

File: daemon/oci_windows.go
Comment (thaJeztah) [reply daemon/oci_windows.go]: > In my experience though, that can fast become a pain with quote escaping when eg building a command line in a bash script, that's why I thought base64 might be a better idea

You mean when using (e.g.) `curl` to create API calls? Or when using a bash script that's using the (`docker`) cli?

If we want to add this as an option on the docker cli, I assume the cli itself would take care of properly encoding the JSON. (perhaps the option should be named `raw` in that case)

Should `stdin` be supported there? e.g.:

```
cat ./my-creds.json | docker service create --credential-specs=file://- ....
```

(using `-` as name for the file to indicate it's from `stdin`)

Comment (wk8) [reply daemon/oci_windows.go]: Fair enough, changing to using raw strings.

**Comment by wk8** [comment]
@thaJeztah : changed to accept raw strings, and fixed the failing test (and added more). PTAL.

File: daemon/oci_windows.go
Comment (cpuguy83) [reply daemon/oci_windows.go]: Can we go ahead and remove config from this?

Comment (wk8) [reply daemon/oci_windows.go]: As we discussed on the maintainers' call yesterday, I'll do that in a follow-up PR.

Comment (cpuguy83) [reply daemon/oci_windows.go]: Why not just in a separate commit?

Comment (cpuguy83) [reply daemon/oci_windows.go]: Seems like the whole reason to add this is to remove config.

Comment (wk8) [reply daemon/oci_windows.go]: Well, the original reason for this is to remove the whole awkward dance I had to do in the kubelet to write/clean up registry keys to add GMSA support there (see the PR message).

And I'll be happy to do that in a follow-up PR; it can't hurt to keep PRs to a reasonable size?

Comment (cpuguy83) [reply daemon/oci_windows.go]: Fine, I would prefer to have the change come in through a single PR but in separate commits but 🤷‍♂️

**Review by cpuguy83 — APPROVED** [review]
LGTM

File: daemon/oci_windows.go
Comment (wk8) [reply daemon/oci_windows.go]: It arguably is 2 different things; I'll make a follow-up PR as soon as this one is merged. Thanks!

**Review by thaJeztah — APPROVED** [review]
LGTM; left a suggestion for using `gotest.tools` for creating the directory and test-file, but let me know if you like that or not 😅

**Comment by wk8** [comment]
@thaJeztah : thanks, made the change you requested. Didn't change the way the file is written though, because that would have pushed the file creation too far from where it's actually used, and made the test more confusing IMO.

File: daemon/oci_windows_test.go
Comment (wk8) [reply daemon/oci_windows_test.go]: Can't do it for the second part, as I do need a specific dir name in there - and moving it all up to the very beginning of the test, far from where it's actually used, makes it a bit confusing IMO.

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by wk8** [comment]
Thanks! :)

**Comment by Drewster727** [comment]
@thaJeztah @wk8 Is there any documentation on this? 
I have a swarm setup with windows nodes and would like to pass credential specs. However, I do not want to add the spec file to each node, which is why I liked @wk8's original thought of passing it with the **service create** and compose.yml directly.

Is this something we can do now on 19.03?

Thanks,
Drew

**Pull request merged**
